auth: Add JWT generation and validation

juliaogris opened the pull request:
Add generated JWT to user struct on successful login and registration.
Create JWT validation method on authenticator.

Misc: Fix CI error for setup-go action.

camh- commented on pkg/foxtrot/jwt.go:
```diff
@@ -0,0 +1,80 @@
+type secret []byte
+
+func (s secret) newJWT(sub string) string {
+	exp := time.Now().AddDate(0, 3, 0).Unix()
```
i assume this is a 3 month expiry? it's a shame you cannot have month durations and have it a parameter. 3 months seems a little random - perhaps a comment saying it is chosen at random-ish?

juliaogris replied:
done.

juliaogris committed:
github: Update setup-go action version

Update setup-go action version from 2.1.2 to 2, in hope to fix a new error on CI:

	Error: Unable to process command '::set-env name=GOROOT::/opt/hostedtoolcache/go/1.15.4/x64' successfully.
	Error: The `set-env` command is disabled. Please upgrade to using Environment Files or opt into unsecure command execution by setting the `ACTIONS_ALLOW_UNSECURE_COMMANDS` environment variable to `true`. For more information see: https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/

juliaogris committed:
auth: Add JWT generation and validation

Add generated JWT to user struct on successful login and registration.
Create JWT validation method on authenticator.

camh- commented on pkg/foxtrot/jwt.go:
```diff
@@ -0,0 +1,73 @@
+func validateJWT(jwt string, secret []byte) error {
+	parts := strings.Split(jwt, ".")
+	if len(parts) != 3 {
+		return errs.Errorf("%v: invalid format, expected two '.' got %d", errJWT, len(parts))
```
should be `len(parts)-1`

juliaogris replied:
Done.

camh- commented on pkg/foxtrot/jwt.go:
```diff
@@ -0,0 +1,73 @@
+	if parts[2] != expectedSignature {
+		return errJWTSignature
+	}
+	if p.Exp < time.Now().Unix() {
```
if `p.Exp` is the expiry time, then this should be `<=` as it expires when it reaches the expiry time, not when it passes it.

https://tools.ietf.org/html/rfc7519#section-4.1.4 says "the expiration time on or after which the JWT MUST NOT be accepted for processing"

juliaogris replied:
Done, thanks.

camh- commented on pkg/foxtrot/jwt.go:
```diff
@@ -0,0 +1,73 @@
+	if err := json.Unmarshal(b, &p); err != nil {
+		return errs.Errorf("%v: json: %v", errJWTEncoding, err)
+	}
+	expectedSignature := sign(encodedJWTHeader+"."+payload, secret)
```
I would check the signature before decoding the payload. This prevents you operating on data provided by an attacker that is not valid. I would also use `parts[0]` here instead of `encodedJWTHeader` because that is what came in from the outside. I know you check it for equality, but in future, that check could change so you would then need to change this anyway.

camh- reviewed (dismissed):
LGTM :avocado:

camh- reviewed (approved):
:avocado:
